k2/csrc/array_ops.cu

```diff
@@ -147,63 +148,55 @@ Array1<int32_t> SpliceRowSplits(int32_t num_arrays,
 
 bool ValidateRowIds(Array1<int32_t> &row_ids, Array1<int32_t> *temp) {
   ContextPtr ctx = row_ids.Context();
-  int32_t *data = row_ids.Data();
+  const int32_t *data = row_ids.Data();
   int32_t dim = row_ids.Dim();
-  if (dim == 0) return true;  // will treat this as valid.a
-
-  if (ctx->GetDeviceType() == kCpu) {
-    if (data[0] < 0) return false;
-    for (int32_t i = 0; i + 1 < dim; i++)
-      if (data[i] > data[i + 1]) return false;
-  }
+  if (dim == 0) return true;  // will treat this as valid
+  // note `row_ids[0]` may copy memory from device to host
+  if (row_ids[0] < 0) return false;
 
   Array1<int32_t> temp_array;
   if (temp == nullptr || temp->Dim() == 0) {
-    temp_array = Array1<int32_t>(row_ids.Context(), 1);
-    temp = &temp_array;
+    temp_array = Array1<int32_t>(ctx, 1);
+  } else {
+    temp_array = temp->Range(0, 1);
```

Comment:
If `temp != nullptr`, we have to check it is compatible with `row_ids`.

If the signature was `bool ValidateRowIds(const Array1<int32_t> &row_ids)`,
the implementation would be much simpler.

Reply:
Well, the purpose we add `temp` parameter here is to save one memory allocation if the caller provide it. There may be some situation that the caller re-use some idle storage as tmp, or some storage has not been freed but will not be used anymore. Another possilbe usage is: we may allocate one extra element for Array1 finally (I mean, change Array1's implementation) to make code easy to use (e.g. the case in ExcusiveSum where dest.Dim() == src.Dim() + 1), so the user even can pass `row_splits.Range(-1, 1)` here. Anyway, we may remove parameter `temp` here if we finally find we would not use it frequently, but for now, let's just keep it.